### ISSUE TITLE
Sort event Attributes

### ISSUE THREAD

**Issue by vieux** [docker/docker#19559]
I updated swarm to use the new events structs with `Attributes` it's really nice, swarm ca now inject a ton a for useful metadata.

I have one comment however, would it be possible to sort the `Attributes` using the keys ?

```bash
$ docker -H :2375 events
2016-01-16T13:57:29.947065777-08:00 network create 465b4e93636028144750917f478d0130fe975c010b28347e4dcb052fb4fcdd5f (name=test2, node.addr=:4242, node.id=XUNN:IXN6:2J5U:C6KX:GQSD:FPPP:B5JZ:CKWV:A6LU:7FQW:MN7W:XS2J, node.ip=<nil>, node.name=dev, type=bridge)
2016-01-16T13:57:43.979931266-08:00 container create d8ae769edd338e6f73a8822a67efda28284ea9c100f26159bd74949db10b562a (node.id=XUNN:IXN6:2J5U:C6KX:GQSD:FPPP:B5JZ:CKWV:A6LU:7FQW:MN7W:XS2J, node.ip=<nil>, node.name=dev, image=busybox, name=stoic_einstein, node.addr=:4242)
2016-01-16T13:57:44.020195568-08:00 network connect af2393d8a130c01bbbc1113cb62dc67f17f372e2b069873e87186f63b4016f9b (node.addr=:4242, node.id=XUNN:IXN6:2J5U:C6KX:GQSD:FPPP:B5JZ:CKWV:A6LU:7FQW:MN7W:XS2J, node.ip=<nil>, node.name=dev, type=bridge, container=d8ae769edd338e6f73a8822a67efda28284ea9c100f26159bd74949db10b562a, name=bridge)
2016-01-16T13:57:44.021229251-08:00 container start d8ae769edd338e6f73a8822a67efda28284ea9c100f26159bd74949db10b562a (node.addr=:4242, node.id=XUNN:IXN6:2J5U:C6KX:GQSD:FPPP:B5JZ:CKWV:A6LU:7FQW:MN7W:XS2J, node.ip=<nil>, node.name=dev, image=busybox, name=stoic_einstein)
2016-01-16T13:58:08.158086191-08:00 network connect 465b4e93636028144750917f478d0130fe975c010b28347e4dcb052fb4fcdd5f (node.ip=<nil>, node.name=dev, type=bridge, container=d8ae769edd338e6f73a8822a67efda28284ea9c100f26159bd74949db10b562a, name=test2, node.addr=:4242, node.id=XUNN:IXN6:2J5U:C6KX:GQSD:FPPP:B5JZ:CKWV:A6LU:7FQW:MN7W:XS2J)
```

@calavera

**Comment by calavera** [comment]
sounds like a good idea.

**Comment by thaJeztah** [comment]
sgtm, would that impact performance-wise or neglectable?

**Comment by calavera** [comment]
> neglectable

sorting always makes things slower, I think we should not care much as long as that sorting happens in the client.

**Comment by calavera** [comment]
to be clear, I don't want to sort in the daemon. It should not care about the attributes, this is a presentation issue that should happen in the consumers/clients.

**Comment by thaJeztah** [comment]
thanks, sounds good!

**Comment by abronan** [comment]
SGTM too

**Comment by HackToday** [comment]
Hi @vieux from https://github.com/docker/docker/blob/master/docs/reference/commandline/events.md
if I understand it right, the events are printed according to time order. 
What do you mean "sort the Attributes using the keys" ? Do you mean use filter or something else ?
And docker events support lots of filters.

Also @calavera what do you mean should happen in the consumers/clients ? Seems docker have related code for events cli

**Comment by vieux** [comment]
@HackToday I'm talking about what's in between the `()` in everyline, sometimes it starts by `node.id`, sometimes with `node.ip` and sometimes with `node.addr`

**Comment by HackToday** [comment]
Thanks @vieux  Got it

**Comment by HackToday** [comment]
Hi @vieux more inputs, if we just sort in client code, it is like this:

```
+               var keys []string
+               var v string
+               for k := range event.Actor.Attributes {
+                       keys = append(keys, k)
+               }
+               sort.Strings(keys)
+               for _, k := range keys {
+                       v = event.Actor.Attributes[k]
                        attrs = append(attrs, fmt.Sprintf("%s=%s", k, v))
                }
                fmt.Fprintf(output, " (%s)", strings.Join(attrs, ", "))
```

It add sort, so a bit slow.


Another way to solve the issue, maybe we check each event type, and output event type according to its format, like below:

if  image type:
   Fprintf(image name,  etc.)
if container type:
   Fprintf(container name,  etc.)

it needs related output standard for that.

**Comment by calavera** [comment]
@HackToday, yes, that would be enough, do you mind opening a PR with the patch?

**Comment by HackToday** [comment]
@calavera  Thanks. OK. I'd like to send a PR.

**Comment by HackToday** [comment]
#dibs

**Comment by HackToday** [comment]
Hi @vieux  Could you check and try this https://github.com/docker/docker/pull/19761 ?
Is it what you needed ?  Thanks

**Comment by LK4D4** [comment]
Closed in #19761
